src/rigid_pendulum_poc/solver.cpp

```diff
@@ -57,4 +54,190 @@ void solve_linear_system(
+)
+    : initial_time_(initial_time), time_step_(time_step), number_of_steps_(number_of_steps) {
+    this->current_time_ = initial_time;
+    this->n_iterations_ = 0;
+    this->total_n_iterations_ = 0;
+}
+
+std::vector<State> GeneralizedAlphaTimeIntegrator::Integrate(const State& initial_state) {
+    auto log = util::Log::Get();
+
+    std::vector<State> states{initial_state};
+    for (size_t i = 0; i < this->number_of_steps_; i++) {
+        log->Debug("Integrating step " + std::to_string(i + 1) + "\n");
+        states.emplace_back(this->AlphaStep(states[i]));
+        this->AdvanceTimeStep();
+    }
+
+    log->Debug("Time integration completed successfully!\n");
+
+    return states;
+}
+
+State GeneralizedAlphaTimeIntegrator::AlphaStep(const State& state) {
+    auto gen_coords = state.GetGeneralizedCoordinates();
+    auto gen_velocity = state.GetGeneralizedVelocity();
+    auto gen_accln = state.GetGeneralizedAcceleration();
+    auto algo_accln = state.GetAlgorithmicAcceleration();
+
+    auto [linear_coords, linear_velocity, linear_algo_accln] =
+        UpdateLinearSolution(gen_coords, gen_velocity, gen_accln, algo_accln);
+
+    gen_coords = linear_coords;
+    gen_velocity = linear_velocity;
+    algo_accln = linear_algo_accln;
+
+    const auto h = this->time_step_;
+    const auto size = gen_coords.size();
+    const double beta_prime = (1 - kALPHA_M) / (h * h * kBETA * (1 - kALPHA_F));
+    const double gamma_prime = kGAMMA / (h * kBETA);
+
+    auto log = util::Log::Get();
+
+    // Perform Newton-Raphson iterations to update nonlinear part of generalized-alpha algorithm
+    for (n_iterations_ = 0; n_iterations_ < kMAX_ITERATIONS; n_iterations_++) {
+        log->Debug("Iteration: " + std::to_string(this->n_iterations_ + 1) + "\n");
+
+        auto residuals = ComputeResiduals(gen_coords);
+        // TODO: Provide actual force increments
+        auto increments = residuals;
+
+        if (this->CheckConvergence(residuals, increments)) {
+            break;
+        }
+
+        auto iteration_matrix = ComputeIterationMatrix(gen_coords);
+        auto gen_coords_delta = residuals;
+        solve_linear_system(iteration_matrix, gen_coords_delta);
+
+        Kokkos::parallel_for(
+            size,
+            KOKKOS_LAMBDA(const int i) {
+                gen_coords(i) += gen_coords_delta(i);
+                gen_velocity(i) += gamma_prime * gen_coords_delta(i);
+                gen_accln(i) += beta_prime * gen_coords_delta(i);
+            }
+        );
+    }
+
+    log->Debug("Converged in " + std::to_string(this->n_iterations_) + " iterations\n");
+    this->total_n_iterations_ += this->n_iterations_;
+
+    Kokkos::parallel_for(
+        size,
+        KOKKOS_LAMBDA(const int i) {
+            // Update algorithmic acceleration once soln has converged
+            algo_accln(i) += (1 - kALPHA_F) / (1 - kALPHA_M) * gen_accln(i);
+        }
+    );
+
+    return State(gen_coords, gen_velocity, gen_accln, algo_accln);
+}
```

Comment:
This structure of this function looks good (assuming that the math is correct).  Let's focus on ComputeResudials and ComputeIterationMatrix as our customization points.  They should eventually call out to a function object that has specific information for how to do this for our problem or simple objects used for testing.  

We may want to be able to simplify the alpha, beta, and gamma values for initial testing (set them to 1 or 0, depending on what makes the math work out simplest).  This should help to show that they are implemented correctly and simplify our initial testing of the AlphaStep function

Reply:
> This structure of this function looks good (assuming that the math is correct). Let's focus on ComputeResudials and ComputeIterationMatrix as our customization points. They should eventually call out to a function object that has specific information for how to do this for our problem or simple objects used for testing.

As discussed above, this sounds like a very reasonable approach to me.

> We may want to be able to simplify the alpha, beta, and gamma values for initial testing (set them to 1 or 0, depending on what makes the math work out simplest). This should help to show that they are implemented correctly and simplify our initial testing of the AlphaStep function

Yes, this is also a reasonable approach. We do need to provide some guard rails around the values of the constants, I will implement that as we get done with the testing part of `AlphaStep()`.